analytics_nudge.md

```diff
@@ -0,0 +1,57 @@
+
+## Proposed Implementation
+Broadly speaking, the components of this feature are:
+1. *when* we nudge, and
+2. *how* we nudge
+    1. how we communicate to the user, "you need to opt in/out?"
+    2. how the user then makes a choice, and how that choice propagates back to the analytics library
+
+### When do we nudge?
+
+We want to nudge users after they've seen some value from Tilt. Possible times to nudge them:
+* when they reach workflow milestone X (e.g., at least one resource is up and running)
+    * probably the easiest of all of these suggestions to verify, since we have the engine state at our disposal
+* Y time after after install
+    * write first_run timestamp to `~/.windmill/tilt` and check against that?
+    * check modified time of the binary?
+* after Z `tilt up` runs
+    * write run_count to `~/windmill/tilt`?
+    
+I propose that we nudge the user **when they get their first resource up and running** (not including k8s_yaml), on the grounds that a. we know that they've seen Tilt *do* something by this point, and b. of all of the above options, this is the easiest to determine with the information we have available (i.e. we just have to check the engine state). 
+    
+### How do we nudge, and how does the user choose?
+
+We should nudge the user in the Web UI, and allow them to select "opt-in" or "opt-out" via button click.
+
+When Tilt determines that we need to nudge the user, we'll send that information along to the Web UI (as part of the Webview json) and render an unobtrusive nudge in sidebar (area B in this image), though in the status bar (C) or in the nav bar (A, though this conflicts with Sail share info) are also possibilities. (Particularly if we nudge when the user gets their first resource up and running, seems safe to assume that we'll have space to spare in the sidebar.)
+
+![possible places in web UI to put analytics nudge](images/web-ui-where-nudge.png) 
+
+The nudge will consist of some copy tbd, and buttons for "opt in" and "opt out". Clicking one of those buttons sends an HTTP request back to the HUD server, which propagates it to the relevant analytics code (which writes the user's choice to disk, as normal). (If something goes wrong with this process, we log the error and replace the buttons with a "something went wrong, try again later" message.)
+
+Copy TDB: either something straight-forward like "opt into analytics to help us better support you!" or a reference to the condition under which we nudged (see above) like, "Congrats on your first Tilt deploy 🎉 want to send analytics to help us better support you?" Will contain a "read more" link to a page on our website about privacy policy, how we use metrics, what metrics we send, etc.
+
+#### Alternatives considered
+1. nudge in the TUI.
+    * we don't want to be adding more functionality to the TUI at this time, as we're trying to move towards the Web UI as the way of the future
```

Comment:
fwiw, this line simply isn't true.

Han and I scratched our heads about this yesterday -- the user feedback has been very strong that we needed a more interactive web UI for investigate/debug user tasks. But neither of us had made a decision either way on the overall direction of the TUI. We were both confused about where people had gotten the impression that we *had* made a decision on this

Reply:
Oh! I was under the impression that we were adding any new features to the web UI first, at least. but I also couldn't tell you where I got that idea.

it IS true that we're focusing on building out the web ui more than we're focusing on building out the TUI at the moment tho, right?

Reply:
Let me try an analogy.

Suppose I asked you, "hey, Maia, let's get lunch. Based on lunch yesterday, it sounds like we're focused on spoons first, and not focusing on forks at the moment"

Like...yes...we used spoons yesterday. And yes, it is true that we've been adding web UI features recently. But if you're focused so much on the tool we're using, rather than the problem we're trying to solve with that tool, I'm worried that you'll make poor decisions.

Reply:
Mmm fair. I think my reasoning does come down to "we're pushing on the web UI right now", but more because that's what we're putting in front of users so that's where we expect (/hope) that new users will spend most of their time. Updated this part of the spec to reflect this.